In [38]:
import bamboolib as bam #easy to do data manipulation and highly recommend going to https://bamboolib.8080labs.com/ and that will help you follow along
import pandas as pd #data manipulation
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

Load File

In [39]:
y=pd.read_csv(r'C:\Users\Trainings\KickstarterData.csv')

Use bamboolib by calling the dataframe which in this case is y. Using the interface we can start investigating the data. Hit show Bamboolib UI.

In [40]:
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0             1    7/1/2019  female           100.0                     white   
1             2    7/1/2019  female           100.0                      blue   
2             3    7/1/2019  female           100.0                    silver   
3             4    7/1/2019  female           100.0                    silver   
4             5    7/1/2019    male           100.0                     white   
...         ...         ...     ...             ...                       ...   
9995       9996    7/9/2019    male           100.0                     black   
9996       9997    7/9/2019    male           100.0                    silver   
9997       9998    7/9/2019    male             NaN                      blue   
9998       9999    7/9/2019    male             NaN                      blue   
9999      10000    7/9/2019    male             NaN                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                      4.0                      vanilla   
1                                      5.0                      vanilla   
2                                      6.0                      vanilla   
3                                      5.0                      vanilla   
4                                      1.0                      vanilla   
...                                    ...                          ...   
9995                                   4.0                    chocolate   
9996                                   6.0                no preference   
9997                                   3.0                    chocolate   
9998                                   6.0                      vanilla   
9999                                   7.0                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

Lets look at missing data by hitting explore dataframe. We see that there is missing data in deposit amount, Ice Cream and Household data. Lets explore how we could replace those. In thise example we will look at deposit amount

In [41]:
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0             1    7/1/2019  female           100.0                     white   
1             2    7/1/2019  female           100.0                      blue   
2             3    7/1/2019  female           100.0                    silver   
3             4    7/1/2019  female           100.0                    silver   
4             5    7/1/2019    male           100.0                     white   
...         ...         ...     ...             ...                       ...   
9995       9996    7/9/2019    male           100.0                     black   
9996       9997    7/9/2019    male           100.0                    silver   
9997       9998    7/9/2019    male             NaN                      blue   
9998       9999    7/9/2019    male             NaN                      blue   
9999      10000    7/9/2019    male             NaN                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                      4.0                      vanilla   
1                                      5.0                      vanilla   
2                                      6.0                      vanilla   
3                                      5.0                      vanilla   
4                                      1.0                      vanilla   
...                                    ...                          ...   
9995                                   4.0                    chocolate   
9996                                   6.0                no preference   
9997                                   3.0                    chocolate   
9998                                   6.0                      vanilla   
9999                                   7.0                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

Looks like the overwhelming value is typically 100, so this would be a case to potentially use the Mode, which would be 100. Also good for spotting outliers too as we spot an abnormal value of 10000.

In [42]:
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0             1    7/1/2019  female           100.0                     white   
1             2    7/1/2019  female           100.0                      blue   
2             3    7/1/2019  female           100.0                    silver   
3             4    7/1/2019  female           100.0                    silver   
4             5    7/1/2019    male           100.0                     white   
...         ...         ...     ...             ...                       ...   
9995       9996    7/9/2019    male           100.0                     black   
9996       9997    7/9/2019    male           100.0                    silver   
9997       9998    7/9/2019    male             NaN                      blue   
9998       9999    7/9/2019    male             NaN                      blue   
9999      10000    7/9/2019    male             NaN                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                      4.0                      vanilla   
1                                      5.0                      vanilla   
2                                      6.0                      vanilla   
3                                      5.0                      vanilla   
4                                      1.0                      vanilla   
...                                    ...                          ...   
9995                                   4.0                    chocolate   
9996                                   6.0                no preference   
9997                                   3.0                    chocolate   
9998                                   6.0                      vanilla   
9999                                   7.0                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

Using the transformations, there is find missing values and replace. You can now either replace with the mode or 100.

In [43]:
y[['Deposit Amount']] = y[['Deposit Amount']].fillna(y[['Deposit Amount']].mode().iloc[0])
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0             1    7/1/2019  female           100.0                     white   
1             2    7/1/2019  female           100.0                      blue   
2             3    7/1/2019  female           100.0                    silver   
3             4    7/1/2019  female           100.0                    silver   
4             5    7/1/2019    male           100.0                     white   
...         ...         ...     ...             ...                       ...   
9995       9996    7/9/2019    male           100.0                     black   
9996       9997    7/9/2019    male           100.0                    silver   
9997       9998    7/9/2019    male           100.0                      blue   
9998       9999    7/9/2019    male           100.0                      blue   
9999      10000    7/9/2019    male           100.0                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                      4.0                      vanilla   
1                                      5.0                      vanilla   
2                                      6.0                      vanilla   
3                                      5.0                      vanilla   
4                                      1.0                      vanilla   
...                                    ...                          ...   
9995                                   4.0                    chocolate   
9996                                   6.0                no preference   
9997                                   3.0                    chocolate   
9998                                   6.0                      vanilla   
9999                                   7.0                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

Now would be where you explore variables to see what flavors are more frequent and if there are any relationships. Using Create Plot will help. This one we see that the preferred color is similar for all donators.

In [44]:
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0             1    7/1/2019  female           100.0                     white   
1             2    7/1/2019  female           100.0                      blue   
2             3    7/1/2019  female           100.0                    silver   
3             4    7/1/2019  female           100.0                    silver   
4             5    7/1/2019    male           100.0                     white   
...         ...         ...     ...             ...                       ...   
9995       9996    7/9/2019    male           100.0                     black   
9996       9997    7/9/2019    male           100.0                    silver   
9997       9998    7/9/2019    male           100.0                      blue   
9998       9999    7/9/2019    male           100.0                      blue   
9999      10000    7/9/2019    male           100.0                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                      4.0                      vanilla   
1                                      5.0                      vanilla   
2                                      6.0                      vanilla   
3                                      5.0                      vanilla   
4                                      1.0                      vanilla   
...                                    ...                          ...   
9995                                   4.0                    chocolate   
9996                                   6.0                no preference   
9997                                   3.0                    chocolate   
9998                                   6.0                      vanilla   
9999                                   7.0                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

Here we can see that on average, those that like Chocolate eat more desserts per week than other flavors.

In [45]:
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0             1    7/1/2019  female           100.0                     white   
1             2    7/1/2019  female           100.0                      blue   
2             3    7/1/2019  female           100.0                    silver   
3             4    7/1/2019  female           100.0                    silver   
4             5    7/1/2019    male           100.0                     white   
...         ...         ...     ...             ...                       ...   
9995       9996    7/9/2019    male           100.0                     black   
9996       9997    7/9/2019    male           100.0                    silver   
9997       9998    7/9/2019    male           100.0                      blue   
9998       9999    7/9/2019    male           100.0                      blue   
9999      10000    7/9/2019    male           100.0                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                      4.0                      vanilla   
1                                      5.0                      vanilla   
2                                      6.0                      vanilla   
3                                      5.0                      vanilla   
4                                      1.0                      vanilla   
...                                    ...                          ...   
9995                                   4.0                    chocolate   
9996                                   6.0                no preference   
9997                                   3.0                    chocolate   
9998                                   6.0                      vanilla   
9999                                   7.0                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

One of the last things we can look at before modeling, is the correlation between variables to address potential overfitting. This can be done below by hitting explore data frame and correlation matrix. Also make sure you convert Donator ID to a string.

In [46]:
y['Donate ID'] = y['Donate ID'].astype('string')
y['Ice Cream Products Consumed Per Week'] = y['Ice Cream Products Consumed Per Week'].astype('Int64')
y[['Ice Cream Products Consumed Per Week']] = y[['Ice Cream Products Consumed Per Week']].fillna(y[['Ice Cream Products Consumed Per Week']].median())
y

Donate ID Donate Date  Gender  Deposit Amount Preferred Color of Device  \
0            1    7/1/2019  female           100.0                     white   
1            2    7/1/2019  female           100.0                      blue   
2            3    7/1/2019  female           100.0                    silver   
3            4    7/1/2019  female           100.0                    silver   
4            5    7/1/2019    male           100.0                     white   
...        ...         ...     ...             ...                       ...   
9995      9996    7/9/2019    male           100.0                     black   
9996      9997    7/9/2019    male           100.0                    silver   
9997      9998    7/9/2019    male           100.0                      blue   
9998      9999    7/9/2019    male           100.0                      blue   
9999     10000    7/9/2019    male           100.0                     black   

      Ice Cream Products Consumed Per Week Favorite Flavor Of Ice Cream  \
0                                        4                      vanilla   
1                                        5                      vanilla   
2                                        6                      vanilla   
3                                        5                      vanilla   
4                                        1                      vanilla   
...                                    ...                          ...   
9995                                     4                    chocolate   
9996                                     6                no preference   
9997                                     3                    chocolate   
9998                                     6                      vanilla   
9999                                     7                    chocolate   

     Donated To Kick Starter Before Household Income Do you own a Keurig  \
0                               yes             <50K                 yes   
1                               yes             <50K                 yes   
2                               yes             <50K                 yes   
3                               yes             <50K                 yes   
4                               yes             <50K                 yes   
...                             ...              ...                 ...   
9995                             no              NaN                  no   
9996                            yes              NaN                 yes   
9997                            yes              NaN                 yes   
9998                             no              NaN                  no   
9999                            yes              NaN                 yes   

      How many desserts do you eat a week  
0                                       7  
1                                       5  
2                                       1  
3                                       4  
4                                       7  
...                                   ...  
9995                                    5  
9996                                    6  
9997                                    5  
9998                                    5  
9999                                    4  

[10000 rows x 11 columns]

Lets try to run a quick KNN Model to see if we can cluster our groups by their device color preference. The first part we will separate the data set into a train and test (so we build the algorithm on one and test the accuracy on the other). In order to to that we need to split our dependent variable from our indepedent variable. We will make x all the independent variables and z the dependent variables. (for this model we will only use numerical variables as we should be coding string variables into dummy variables for a knn model. as you will see there is an option for one hot encoding which will make your life easier to do that.)

In [47]:
y
x = y[['Deposit Amount', 'Ice Cream Products Consumed Per Week', 'How many desserts do you eat a week']]
z = y[['Preferred Color of Device']]
z

Preferred Color of Device
0                        white
1                         blue
2                       silver
3                       silver
4                        white
...                        ...
9995                     black
9996                    silver
9997                      blue
9998                      blue
9999                     black

[10000 rows x 1 columns]

Here is the data set split code and model.You can test out varying n_neighbor values but for now I am just starting with 5

In [48]:
from sklearn.model_selection import train_test_split
x_train, x_test, z_train, z_test = train_test_split(x, z, test_size=0.20)

from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(x_train, z_train)

KNeighborsClassifier()

Now we can look at our results. The first part of the code gets us predictions while second part gets us our confusion matrix

In [49]:
z_pred = classifier.predict(x_test)
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(z_test, z_pred))
print(classification_report(z_test, z_pred))

[[89 60 48 49 39 46]
 [97 63 55 44 40 44]
 [94 62 45 36 47 46]
 [95 55 36 43 37 43]
 [99 57 50 52 50 48]
 [78 69 42 45 50 47]]
               precision    recall  f1-score   support

        black       0.16      0.27      0.20       331
         blue       0.17      0.18      0.18       343
no preference       0.16      0.14      0.15       330
          red       0.16      0.14      0.15       309
       silver       0.19      0.14      0.16       356
        white       0.17      0.14      0.16       331

     accuracy                           0.17      2000
    macro avg       0.17      0.17      0.17      2000
 weighted avg       0.17      0.17      0.17      2000



We can also visualizae the results as well

Overall we can this model isn't very good as the accuracy is relatively low (~17%) and precision and recall is weak on all the colors.